Optional `adduser()` method for auth plug-ins is only partly optional

<!-- PLEASE READ THIS:  
 - If you are not sure is a bug, OPEN a DISCUSSION, if is a legitimate bug, is easy to create a bug from a discussion.
 - Empty reports won't be considered and eventually be closed by a bot.
 - Include debugging notes will help to fix it faster, HOW TO: https://github.com/verdaccio/verdaccio/wiki/Debugging-Verdaccio 
 - If you remove this template, ticket will be closed immediately.
 - No English perfect is required, use public translators if is need it, we will do our best to help you.
 - Extra bonus: The most complete this report is delivered, the faster you will get a response.
 - Extra bonus: include screenshots, logs (remove sensitive data).
 - If you are willing to fix it, there is a checkbox at the bottom.
-->

**Your Environment**
  <!-- bug below the version 5.x will be closed, see SECURITY.md for more details -->
 * **verdaccio version**: 5.23.2
 * **node version** : v18.15.0
 * **package manager**: 9.5.0
 * **os**: Linux
 * **platform**: npm

**Describe the bug**

I've been writing a basic plug-in for a Verdaccio server I'm working on using [the API described here](https://verdaccio.org/docs/plugin-auth). The interface and documentation suggests that it is optional, and while Verdaccio does run without it, `npm login` will always fail.


**To Reproduce**

Create a simple plug-in that doesn't have an `adduser()` function:

```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });

const createError = require('http-errors');
const commonsApi = require('@verdaccio/commons-api')

/**
 * Editor custom authentication plug-in
 */
class AuthCustomPlugin {
	constructor(config, options) {
		this.logger = options.logger;

		return this;
	}

	authenticate(user, password, cb) {
		cb(null, []);
	}

	allow_access(user, pkg, cb) {
		cb(null, true);
	}

	allow_publish(user, pkg, cb) {
		cb(null, true);
	}

	allow_unpublish(user, pkg, cb) {
		cb(null, true);
	}
}
exports.default = AuthCustomPlugin;
```

Run Verdaccio with this plug-in and attempt `npm login` for an _existing_ user.  This results in:

```
npm ERR! code E409
npm ERR! 409 Conflict - PUT https://..../-/user/org.couchdb.user:allan/-rev/undefined - username is already registered

npm ERR! A complete log of this run can be found in: /....npm/_logs/2023-04-14T14_40_06_339Z-debug-0.log
```

The console output from Verdaccio is:

```
0|verdaccio  | info <-- 127.0.0.1 requested 'POST /-/v1/login'
0|verdaccio  | http <-- 200, user: null(.... via 127.0.0.1), req: 'POST /-/v1/login', bytes: 22/0
0|verdaccio  | http <-- 404, user: null(.... via 127.0.0.1), req: 'POST /-/v1/login', bytes: 22/150
0|verdaccio  | info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:allan'
0|verdaccio  | http <-- 200, user: null(.... via 127.0.0.1), req: 'PUT /-/user/org.couchdb.user:allan', bytes: 143/0
0|verdaccio  | warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
0|verdaccio  | error--- the user allan could not being added. Error: username is already registered
0|verdaccio  | http <-- 409, user: null(.... via 127.0.0.1), req: 'PUT /-/user/org.couchdb.user:allan', error: username is already registered
0|verdaccio  | info <-- 127.0.0.1 requested 'GET /-/user/org.couchdb.user:allan?write=true'
0|verdaccio  | http <-- 200, user: null(.... via 127.0.0.1), req: 'GET /-/user/org.couchdb.user:allan?write=true', bytes: 0/0
0|verdaccio  | http <-- 304, user: null(.... via 127.0.0.1), req: 'GET /-/user/org.couchdb.user:allan?write=true', bytes: 0/0
0|verdaccio  | info <-- 127.0.0.1 requested 'PUT /-/user/org.couchdb.user:allan/-rev/undefined'
0|verdaccio  | http <-- 200, user: null(.... via 127.0.0.1), req: 'PUT /-/user/org.couchdb.user:allan/-rev/undefined', bytes: 187/0
0|verdaccio  | warn --- the plugin method add_user in the auth plugin is deprecated and will be removed in next major release, notify to the plugin author
0|verdaccio  | error--- the user allan could not being added. Error: username is already registered
0|verdaccio  | http <-- 409, user: null(.... via 127.0.0.1), req: 'PUT /-/user/org.couchdb.user:allan/-rev/undefined', error: username is already registered
```

If you now add the following function to the plug-in:

```
        adduser(user, pass, cb) {
                cb(null, true);
        }
```

Then you can login successfully.

Related: I'd been looking at [this issue](https://github.com/verdaccio/verdaccio/issues/943) before discovering that `adduser()` is required to allow existing users to login.

I'm not sure what the right fix it - either to make it a required method, or to allow `adduser` to pass for the plug-in if it wasn't defined (could anyone be depending on this behaviour?!).

--- a/.changeset/green-eagles-boil.md
+++ b/.changeset/green-eagles-boil.md
@@ -1,0 +1,6 @@
+---
+'@verdaccio/auth': patch
+'@verdaccio/api': patch
+---
+
+fix: E409 username is already registered (adduser)

--- a/packages/api/src/user.ts
+++ b/packages/api/src/user.ts
@@ -28,7 +28,11 @@ export default function (route: Router, auth: Auth, config: Config, logger: Logg
     function (req: $RequestExtend, res: Response, next: $NextFunctionVer): void {
       debug('verifying user');
 
-      if (typeof req.remote_user.name !== 'string' || req.remote_user.name === '') {
+      if (
+        !req.remote_user ||
+        typeof req.remote_user.name !== 'string' ||
+        req.remote_user.name === ''
+      ) {
         debug('user not logged in');
         res.status(HTTP_STATUS.OK);
         return next({ ok: false });

--- a/packages/auth/src/utils.ts
+++ b/packages/auth/src/utils.ts
@@ -166,7 +166,9 @@ export function getDefaultPlugins(logger: Logger): pluginUtils.Auth<Config> {
 
     adduser(_user: string, _password: string, cb: pluginUtils.AuthUserCallback): void {
       debug('triggered default adduser method');
-      return cb(errorUtils.getConflict(API_ERROR.BAD_USERNAME_PASSWORD));
+      // since adduser is not implemented but optional, continue without error
+      // this assumes that the user is added by an external system
+      cb(null, true);
     },
 
     // @ts-ignore

--- a/packages/auth/test/auth-utils.spec.ts
+++ b/packages/auth/test/auth-utils.spec.ts
@@ -123,11 +123,12 @@ describe('Auth utilities', () => {
       });
     });
 
-    test('add user should fail by default (default)', () => {
+    test('add user should not fail by default (default)', () => {
       const plugin = getDefaultPlugins({ trace: vi.fn() });
       // @ts-ignore
-      plugin.adduser('foo', 'bar', (error: any) => {
-        expect(error).toEqual(errorUtils.getForbidden(API_ERROR.BAD_USERNAME_PASSWORD));
+      plugin.adduser('foo', 'bar', (error: any, access: any) => {
+        expect(error).toEqual(null);
+        expect(access).toEqual(true);
       });
     });
   });

--- a/packages/auth/test/auth.spec.ts
+++ b/packages/auth/test/auth.spec.ts
@@ -4,14 +4,7 @@ import supertest from 'supertest';
 import { describe, expect, test, vi } from 'vitest';
 
 import { Config as AppConfig, ROLES, createRemoteUser, getDefaultConfig } from '@verdaccio/config';
-import {
-  API_ERROR,
-  HEADERS,
-  HTTP_STATUS,
-  SUPPORT_ERRORS,
-  TOKEN_BEARER,
-  errorUtils,
-} from '@verdaccio/core';
+import { HEADERS, HTTP_STATUS, SUPPORT_ERRORS, TOKEN_BEARER, errorUtils } from '@verdaccio/core';
 import { logger, setup } from '@verdaccio/logger';
 import { errorReportingMiddleware, final, handleError } from '@verdaccio/middleware';
 import { Config } from '@verdaccio/types';
@@ -259,7 +252,7 @@ describe('AuthTest', () => {
     describe('no custom allow_access implementation provided', () => {
       // when allow_access is not implemented, the groups must match
       // exactly with the packages access group
-      test('should fails if groups do not match exactly', async () => {
+      test('should fail if groups do not match exactly', async () => {
         const config: Config = new AppConfig({ ...authProfileConf });
         config.checkSecretKey('12345');
         const auth: Auth = new Auth(config, logger);
@@ -308,7 +301,7 @@ describe('AuthTest', () => {
     describe('no custom allow_publish implementation provided', () => {
       // when allow_access is not implemented, the groups must match
       // exactly with the packages access group
-      test('should fails if groups do not match exactly', async () => {
+      test('should fail if groups do not match exactly', async () => {
         const config: Config = new AppConfig({ ...authProfileConf });
         config.checkSecretKey('12345');
         const auth: Auth = new Auth(config, logger);
@@ -354,7 +347,7 @@ describe('AuthTest', () => {
   });
   describe('allow_unpublish', () => {
     describe('no custom allow_unpublish implementation provided', () => {
-      test('should fails if groups do not match exactly', async () => {
+      test('should fail if groups do not match exactly', async () => {
         const config: Config = new AppConfig({ ...authProfileConf });
         config.checkSecretKey('12345');
 
@@ -439,7 +432,7 @@ describe('AuthTest', () => {
     describe('error handling', () => {
       // when allow_access is not implemented, the groups must match
       // exactly with the packages access group
-      test('should fails with bad password if adduser is not implemented', async () => {
+      test('should not fail if adduser is not implemented', async () => {
         const config: Config = new AppConfig({ ...authProfileConf });
         config.checkSecretKey('12345');
         const auth: Auth = new Auth(config, logger);
@@ -451,12 +444,21 @@ describe('AuthTest', () => {
         auth.add_user('juan', 'password', callback);
 
         expect(callback).toHaveBeenCalledTimes(1);
-        expect(callback).toHaveBeenCalledWith(
-          errorUtils.getConflict(API_ERROR.BAD_USERNAME_PASSWORD)
-        );
+        expect(callback).toHaveBeenCalledWith(null, {
+          groups: [
+            'test',
+            ROLES.$ALL,
+            ROLES.$AUTH,
+            ROLES.DEPRECATED_ALL,
+            ROLES.DEPRECATED_AUTH,
+            ROLES.ALL,
+          ],
+          name: 'juan',
+          real_groups: ['test'],
+        });
       });
 
-      test('should fails if adduser fails internally (exception)', async () => {
+      test('should fail if adduser fails internally (exception)', async () => {
         const config: Config = new AppConfig({
           ...getDefaultConfig(),
           plugins: path.join(__dirname, './partials/plugin'),
@@ -471,14 +473,14 @@ describe('AuthTest', () => {
 
         const callback = vi.fn();
 
-        // note: fail uas username make plugin fails
+        // note: username to make plugin fail
         auth.add_user('fail', 'password', callback);
 
         expect(callback).toHaveBeenCalledTimes(1);
         expect(callback).toHaveBeenCalledWith(new Error('bad username'));
       });
 
-      test('should skip to the next plugin and fails', async () => {
+      test('should skip to the next plugin and fail', async () => {
         const config: Config = new AppConfig({
           ...getDefaultConfig(),
           plugins: path.join(__dirname, './partials/plugin'),
@@ -495,13 +497,22 @@ describe('AuthTest', () => {
 
         const callback = vi.fn();
 
-        // note: fail uas username make plugin fails
+        // note: username to make plugin fail
         auth.add_user('skip', 'password', callback);
 
         expect(callback).toHaveBeenCalledTimes(1);
-        expect(callback).toHaveBeenCalledWith(
-          errorUtils.getConflict(API_ERROR.BAD_USERNAME_PASSWORD)
-        );
+        expect(callback).toHaveBeenCalledWith(null, {
+          groups: [
+            'test',
+            ROLES.$ALL,
+            ROLES.$AUTH,
+            ROLES.DEPRECATED_ALL,
+            ROLES.DEPRECATED_AUTH,
+            ROLES.ALL,
+          ],
+          name: 'skip',
+          real_groups: ['test'],
+        });
       });
     });
     test('should success if adduser', async () => {

